File: backend/app/schemas/medical.py
"""Pydantic request/response schemas (typed API contracts)."""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


# --- Study ---
class SeriesOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    series_uid: str
    series_number: int | None = None
    description: str | None = None
    modality: str | None = None
    rows: int | None = None
    columns: int | None = None
    slice_count: int
    pixel_spacing_x: float | None = None
    pixel_spacing_y: float | None = None
    slice_thickness: float | None = None
    slice_spacing: float | None = None
    image_orientation: str | None = None
    image_position: str | None = None
    is_selected: bool = False
    warnings: str | None = None


class StudyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    study_uid: str
    name: str
    modality: str | None = None
    status: str
    slice_count: int
    series_count: int
    study_date: str | None = None
    created_at: Any
    updated_at: Any
    error_message: str | None = None


class StudyListOut(BaseModel):
    studies: list[StudyOut]
    total: int


class StudyDetailOut(StudyOut):
    series: list[SeriesOut] = []


# --- Upload ---
class UploadResult(BaseModel):
    study: StudyOut
    series_detected: int
    warnings: list[str] = []


# --- Slice ---
class SliceMetaOut(BaseModel):
    index: int
    total: int
    instance_number: int | None = None
    slice_position: list[float] | None = None
    rows: int
    columns: int
    hu_min: float | None = None
    hu_max: float | None = None
    window_width: float | None = None
    window_level: float | None = None
    rescale_slope: float | None = None
    rescale_intercept: float | None = None


# --- Metadata ---
class MetadataOut(BaseModel):
    study_id: int
    study_uid: str
    patient_id: str | None = None
    modality: str | None = None
    study_date: str | None = None
    series_description: str | None = None
    rows: int | None = None
    columns: int | None = None
    slice_count: int
    pixel_spacing: list[float] | None = None
    slice_thickness: float | None = None
    slice_spacing: float | None = None
    image_orientation: list[float] | None = None
    image_position: list[float] | None = None
    rescale_slope: float | None = None
    rescale_intercept: float | None = None
    physical_size: list[float] | None = None
    warnings: list[str] = []


# --- Reconstruction ---
class ReconstructRequest(BaseModel):
    threshold_hu: float = Field(default=300.0, ge=-2000, le=4000)
    series_id: int | None = None
    method: Literal["marching_cubes"] = "marching_cubes"
    remove_small_components: bool = True
    min_component_fraction: float = Field(default=0.001, ge=0.0, le=1.0)
    smoothing_iterations: int = Field(default=0, ge=0, le=20)


class ReconstructResponse(BaseModel):
    job_id: str


class ModelOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    study_id: int
    series_id: int | None = None
    name: str
    method: str
    threshold_hu: float
    vertices: int
    triangles: int
    bbox_min: list[float] | None = None
    bbox_max: list[float] | None = None
    physical_size: list[float] | None = None
    status: str
    created_at: Any

    @model_validator(mode="before")
    @classmethod
    def _derive_lists(cls, obj: Any) -> Any:
        """Derive list-shaped fields from ORM scalar x/y/z columns."""
        if isinstance(obj, dict):
            return obj
        for prefix, target in ("bbox_min", "bbox_min"), ("bbox_max", "bbox_max"), ("physical_size", "physical_size"):
            vals = (
                getattr(obj, f"{prefix}_x", None),
                getattr(obj, f"{prefix}_y", None),
                getattr(obj, f"{prefix}_z", None),
            )
            if all(v is not None for v in vals):
                setattr(obj, target, [float(v) for v in vals])
        return obj


class ModelListOut(BaseModel):
    models: list[ModelOut]


# --- Measurements ---
class MeasurementRequest(BaseModel):
    points: list[list[float]] = Field(min_length=2)
    measurement_type: Literal["distance", "angle"] = "distance"


class MeasurementOut(BaseModel):
    id: int
    model_id: int
    measurement_type: str
    value: float
    unit: str
    points: list[list[float]]
    created_at: Any


class MeasurementResult(BaseModel):
    measurement: MeasurementOut


# --- Jobs ---
class JobOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    study_id: int
    task_type: str
    status: str
    progress: float
    message: str | None = None
    error: str | None = None
    result_model_id: int | None = None
    created_at: Any
    updated_at: Any


# --- Common error ---
class ErrorResponse(BaseModel):
    detail: str
    code: str | None = None


# --- Dashboard ---
class DashboardStats(BaseModel):
    total_studies: int
    processed_studies: int
    models_generated: int
    recent_studies: list[StudyOut]
    processing_jobs: list[JobOut]

File: frontend/src/types/medical.ts
/** Shared medical/imaging type definitions matching the backend Pydantic schemas. */

export type StudyStatus = "uploaded" | "processing" | "ready" | "failed";
export type JobStatus = "queued" | "processing" | "completed" | "failed";

export interface Study {
  id: number;
  study_uid: string;
  name: string;
  modality: string | null;
  status: StudyStatus;
  slice_count: number;
  series_count: number;
  study_date: string | null;
  created_at: string;
  updated_at: string;
  error_message: string | null;
}

export interface SeriesInfo {
  id: number;
  series_uid: string;
  series_number: number | null;
  description: string | null;
  modality: string | null;
  rows: number | null;
  columns: number | null;
  slice_count: number;
  pixel_spacing_x: number | null;
  pixel_spacing_y: number | null;
  slice_thickness: number | null;
  slice_spacing: number | null;
  image_orientation: string | null;
  image_position: string | null;
  is_selected: boolean;
  warnings: string | null;
}

export interface StudyDetail extends Study {
  series: SeriesInfo[];
}

export interface UploadResult {
  study: Study;
  series_detected: number;
  warnings: string[];
}

export interface Metadata {
  study_id: number;
  study_uid: string;
  patient_id: string | null;
  modality: string | null;
  study_date: string | null;
  series_description: string | null;
  rows: number | null;
  columns: number | null;
  slice_count: number;
  pixel_spacing: number[] | null;
  slice_thickness: number | null;
  slice_spacing: number | null;
  image_orientation: number[] | null;
  image_position: number[] | null;
  rescale_slope: number | null;
  rescale_intercept: number | null;
  physical_size: number[] | null;
  warnings: string[];
}

export interface SliceInfo {
  index: number;
  total: number;
  rows: number;
  columns: number;
  hu_min: number | null;
  hu_max: number | null;
}

export interface SliceResponse extends SliceInfo {
  width: number;
  level: number;
  image_base64: string;
}

export interface Model {
  id: number;
  study_id: number;
  series_id: number | null;
  name: string;
  method: string;
  threshold_hu: number;
  vertices: number;
  triangles: number;
  bbox_min: number[] | null;
  bbox_max: number[] | null;
  physical_size: number[] | null;
  status: string;
  created_at: string;
}

export interface Job {
  id: string;
  study_id: number;
  task_type: string;
  status: JobStatus;
  progress: number;
  message: string | null;
  error: string | null;
  result_model_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface Measurement {
  id: number;
  model_id: number;
  measurement_type: "distance" | "angle";
  value: number;
  unit: string;
  points: number[][];
  created_at: string;
}

export interface DashboardStats {
  total_studies: number;
  processed_studies: number;
  models_generated: number;
  recent_studies: Study[];
  processing_jobs: Job[];
}

export interface ReconstructRequest {
  threshold_hu: number;
  series_id?: number | null;
  method?: "marching_cubes";
  remove_small_components?: boolean;
  min_component_fraction?: number;
  smoothing_iterations?: number;
}

export interface ErrorResponse {
  detail: string;
  code: string | null;
}

// --- Phase 3: AI Segmentation ---
export interface SegmentationModelInfo {
  id: string;
  name: string;
  version: string;
  task: string;
  labels: string[];
  framework: string;
  target_spacing: number[] | null;
  checkpoint_path: string | null;
  checkpoint_available: boolean;
}

export interface SegmentationJobResponse {
  job_id: string;
  status: string;
}

export interface SegmentationLabelOut {
  id: number;
  name: string;
  color: string;
  voxel_count: number;
  volume_cm3: number;
  bbox_min: number[] | null;
  bbox_max: number[] | null;
  centroid: number[] | null;
}

export interface SegmentationResult {
  id: number;
  study_id: number;
  model_id: string;
  model_version: string;
  status: string;
  created_at: string;
  processing_duration_sec: number | null;
  device: string | null;
  labels: SegmentationLabelOut[];
  error: string | null;
}

export interface SegmentationJobStatus {
  id: string;
  study_id: number;
  task_type: string;
  status: string;
  progress: number;
  message: string | null;
  error: string | null;
  result_id: number | null;
}

// --- Phase 2: MPR ---
export type PlaneName = "axial" | "coronal" | "sagittal";

export interface VolumeInfo {
  shape: [number, number, number]; // depth, height, width
  dimensions: [number, number, number]; // width, height, depth
  spacing: [number, number, number]; // sx, sy, sz
  origin: [number, number, number];
  direction: number[]; // 9-element row-major
  hu_min: number;
  hu_max: number;
}

export interface MprSlice {
  plane: PlaneName;
  index: number;
  total: number;
  width: number;
  level: number;
  rows: number;
  columns: number;
  physical_width: number;
  physical_height: number;
  image_base64: string;
}

export interface CoordinateInfo {
  voxel: [number, number, number]; // fractional x,y,z
  index: [number, number, number]; // integer x,y,z = column,row,slice
  clamped_index: [number, number, number];
  world: [number, number, number];
  hu: number;
  in_bounds: boolean;
}

export interface VoxelInfo {
  index: [number, number, number];
  world: [number, number, number];
  hu: number;
}
